File: src/cart/cart.service.ts
import { Injectable, NotFoundException } from "@nestjs/common"
import { CartItem } from "../entities/cart-item.entity"
import { Product } from "../entities/product.entity"
import { User } from "../entities/user.entity"
import { AddToCartDto } from "./dto/add-to-cart.dto"
import { UpdateCartItemDto } from "./dto/update-cart-item.dto"
import { DualDatabaseService } from "../database/dual-database.service"

@Injectable()
export class CartService {
  constructor(private dualDatabaseService: DualDatabaseService) {}

  private async getUserIdFromUsername(username: string): Promise<string> {
    if (username === "anonymous") {
      return "anonymous"
    }

    const user = await this.dualDatabaseService.findOne(username, User, {
      where: { username },
    })

    if (!user) {
      throw new NotFoundException("User not found")
    }

    return user.id
  }

  async addToCart(
    username: string,
    addToCartDto: AddToCartDto
  ): Promise<CartItem> {
    const { product_id, quantity } = addToCartDto

    // Resolve username to user ID for business logic
    const userId = await this.getUserIdFromUsername(username)

    // Check if product exists (use username for feature flag targeting)
    const product = await this.dualDatabaseService.findOne(username, Product, {
      where: { id: product_id },
    })
    if (!product) {
      throw new NotFoundException("Product not found")
    }

    // Check if item already exists in cart (use username for feature flag targeting)
    const existingCartItem = await this.dualDatabaseService.findOne(
      username,
      CartItem,
      {
        where: { user_id: userId, product_id },
        relations: ["product"],
      }
    )

    if (existingCartItem) {
      // Update quantity
      const updatedQuantity = existingCartItem.quantity + quantity
      await this.dualDatabaseService.dualUpdate(CartItem, existingCartItem.id, {
        quantity: updatedQuantity,
      })
      return await this.dualDatabaseService.findOne(username, CartItem, {
        where: { id: existingCartItem.id },
        relations: ["product"],
      })
    } else {
      // Create new cart item
      const cartItem = new CartItem()
      Object.assign(cartItem, {
        user_id: userId,
        product_id,
        quantity,
        product,
      })
      return await this.dualDatabaseService.dualSave(CartItem, cartItem)
    }
  }

  async getCartItems(username: string): Promise<CartItem[]> {
    // Resolve username to user ID for business logic
    const userId = await this.getUserIdFromUsername(username)

    return await this.dualDatabaseService.findMany(username, CartItem, {
      where: { user_id: userId },
      relations: ["product"],
    })
  }

  async updateCartItem(
    username: string,
    id: string,
    updateCartItemDto: UpdateCartItemDto
  ): Promise<CartItem> {
    // Resolve username to user ID for business logic
    const userId = await this.getUserIdFromUsername(username)

    const cartItem = await this.dualDatabaseService.findOne(
      username,
      CartItem,
      {
        where: { id, user_id: userId },
        relations: ["product"],
      }
    )

    if (!cartItem) {
      throw new NotFoundException("Cart item not found")
    }

    await this.dualDatabaseService.dualUpdate(CartItem, id, updateCartItemDto)
    return await this.dualDatabaseService.findOne(username, CartItem, {
      where: { id },
      relations: ["product"],
    })
  }

  async removeCartItem(username: string, id: string): Promise<void> {
    // Resolve username to user ID for business logic
    const userId = await this.getUserIdFromUsername(username)

    const cartItem = await this.dualDatabaseService.findOne(
      username,
      CartItem,
      {
        where: { id, user_id: userId },
      }
    )

    if (!cartItem) {
      throw new NotFoundException("Cart item not found")
    }

    await this.dualDatabaseService.dualDelete(CartItem, id)
  }

  async clearCart(username: string): Promise<void> {
    // Resolve username to user ID for business logic
    const userId = await this.getUserIdFromUsername(username)

    const cartItems = await this.dualDatabaseService.findMany(
      username,
      CartItem,
      {
        where: { user_id: userId },
      }
    )

    for (const item of cartItems) {
      await this.dualDatabaseService.dualDelete(CartItem, item.id)
    }
  }
}

File: java-backend/src/main/java/com/shopper/backend/service/CartService.java
package com.shopper.backend.service;

import com.shopper.backend.dto.AddToCartDto;
import com.shopper.backend.dto.UpdateCartItemDto;
import com.shopper.backend.entity.CartItem;
import com.shopper.backend.entity.Product;
import com.shopper.backend.entity.User;
import com.shopper.backend.repository.CartItemRepository;
import com.shopper.backend.repository.ProductRepository;
import com.shopper.backend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@Transactional
public class CartService {

    @Autowired
    private CartItemRepository cartItemRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private FeatureFlagService featureFlagService;

    public List<CartItem> getCartItems(String username) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // Check feature flag for cart improvements
        boolean cartImprovements = featureFlagService.getBooleanFeatureFlag(
                username, "cart-improvements", false);

        List<CartItem> cartItems = cartItemRepository.findByUserIdWithProduct(user.getId());

        if (cartImprovements) {
            System.out.println("✅ Cart improvements enabled for user: " + username);
        }

        return cartItems;
    }

    public CartItem addToCart(String username, AddToCartDto addToCartDto) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Product product = productRepository.findById(addToCartDto.getProductId())
                .orElseThrow(() -> new RuntimeException("Product not found"));

        // Check if product is in stock
        if (product.getStockQuantity() < addToCartDto.getQuantity()) {
            throw new RuntimeException("Insufficient stock");
        }

        // Check if item already exists in cart
        Optional<CartItem> existingCartItem = cartItemRepository.findByUserIdAndProductId(
                user.getId(), addToCartDto.getProductId());

        if (existingCartItem.isPresent()) {
            // Update existing item
            CartItem cartItem = existingCartItem.get();
            int newQuantity = cartItem.getQuantity() + addToCartDto.getQuantity();
            
            if (product.getStockQuantity() < newQuantity) {
                throw new RuntimeException("Insufficient stock for requested quantity");
            }
            
            cartItem.setQuantity(newQuantity);
            return cartItemRepository.save(cartItem);
        } else {
            // Create new cart item
            CartItem cartItem = new CartItem();
            cartItem.setUserId(user.getId());
            cartItem.setProductId(addToCartDto.getProductId());
            cartItem.setQuantity(addToCartDto.getQuantity());
            return cartItemRepository.save(cartItem);
        }
    }

    public CartItem updateCartItem(String username, UUID cartItemId, UpdateCartItemDto updateCartItemDto) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        CartItem cartItem = cartItemRepository.findByIdAndUserId(cartItemId, user.getId())
                .orElseThrow(() -> new RuntimeException("Cart item not found"));

        Product product = productRepository.findById(cartItem.getProductId())
                .orElseThrow(() -> new RuntimeException("Product not found"));

        // Check if product is in stock
        if (product.getStockQuantity() < updateCartItemDto.getQuantity()) {
            throw new RuntimeException("Insufficient stock");
        }

        cartItem.setQuantity(updateCartItemDto.getQuantity());
        return cartItemRepository.save(cartItem);
    }

    public void removeCartItem(String username, UUID cartItemId) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        CartItem cartItem = cartItemRepository.findByIdAndUserId(cartItemId, user.getId())
                .orElseThrow(() -> new RuntimeException("Cart item not found"));

        cartItemRepository.delete(cartItem);
    }

    public void clearCart(String username) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        cartItemRepository.deleteByUserId(user.getId());
    }

    public Long getCartItemCount(String username) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        return cartItemRepository.countByUserId(user.getId());
    }
}